staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go

```diff
@@ -1664,3 +1673,213 @@ func TestRatchetingFunctionality(t *testing.T) {
+			if err := apiextensionsv1.Convert_v1_CustomResourceValidation_To_apiextensions_CustomResourceValidation(v.Schema, internalValidation, nil); err != nil {
+				b.Fatal(fmt.Errorf("failed converting CRD validation to internal version: %v", err))
+				return
+			}
+
+			validator, err := NewValidator(internalValidation.OpenAPIV3Schema, gvk, parsed.Spec.Scope == apiextensionsv1.NamespaceScoped)
+			if err != nil {
+				b.Fatal(err)
+				return
+			}
+			validators[gvk] = validator
+		}
+
+	}
+
+	// Organize all the files by GVK.
+	gvksToValidFiles := map[schema.GroupKind][]*unstructured.Unstructured{}
+	gvksToInvalidFiles := map[schema.GroupKind][]*unstructured.Unstructured{}
+
+	for _, valid := range validFiles {
+		gvk := valid.GroupVersionKind()
+		gvksToValidFiles[gvk.GroupKind()] = append(gvksToValidFiles[gvk.GroupKind()], valid)
+	}
+
+	for _, invalid := range invalidFiles {
+		gvk := invalid.GroupVersionKind()
+		gvksToInvalidFiles[gvk.GroupKind()] = append(gvksToInvalidFiles[gvk.GroupKind()], invalid)
+	}
+
+	// Remove any GVKs for which we dont have both valid and invalid files.
+	for gvk := range gvksToValidFiles {
+		if _, ok := gvksToInvalidFiles[gvk]; !ok {
+			delete(gvksToValidFiles, gvk)
+		}
+	}
+
+	for gvk := range gvksToInvalidFiles {
+		if _, ok := gvksToValidFiles[gvk]; !ok {
+			delete(gvksToInvalidFiles, gvk)
+		}
+	}
+
+	// Give all objects the same name.
+	for _, files := range gvksToValidFiles {
+		for _, file := range files {
+			file.SetName("mycrd")
+		}
+	}
+	for _, files := range gvksToInvalidFiles {
+		for _, file := range files {
+			file.SetName("mycrd")
+		}
+	}
+
+	// For each valid file, match it with every invalid file of the same GVK
+	pairs := []struct {
+		old *unstructured.Unstructured
+		new *unstructured.Unstructured
+	}{}
+
+	for gvk, valids := range gvksToValidFiles {
+		for _, valid := range valids {
+			for _, invalid := range gvksToInvalidFiles[gvk] {
+				pairs = append(pairs, struct {
+					old *unstructured.Unstructured
+					new *unstructured.Unstructured
+				}{old: valid, new: invalid})
+			}
+		}
+	}
+
+	// For each invalid file, add pair with every other invalid file of the same
+	// GVK including itself
+	for gvk, invalids := range gvksToInvalidFiles {
+		for _, invalid := range invalids {
+			for _, invalid2 := range gvksToInvalidFiles[gvk] {
+				pairs = append(pairs, struct {
+					old *unstructured.Unstructured
+					new *unstructured.Unstructured
+				}{old: invalid, new: invalid2})
+			}
+		}
+	}
+
+	// For each pair, run the ratcheting algorithm on the update.
+	//
+	for _, ratchetingEnabled := range []bool{true, false} {
+		name := "RatchetingEnabled"
+		if !ratchetingEnabled {
+			name = "RatchetingDisabled"
+		}
+		b.Run(name, func(b *testing.B) {
+			defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, features.CRDValidationRatcheting, ratchetingEnabled)()
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				for _, pair := range pairs {
```

Comment:
According to godoc

> A benchmark that calls Run at least once will not be measured itself and will be called once with N=1.

So adding the sub-benchmark at this point I don't think would have the desired effect. I did add one and the text takes forever since it runs each file 10000 times and there are many pairs. 

Instead, I could run each pair in a loop that validates it a preset number of itmes? I can also split the benchmarks up into groups of "no errors to ratchet" and "has errors to try to ratchet"

Reply:
I should also note this benchmark only tests the best case and worst case scenarios. The average cause would be that the comparison is localized to a small portion of an object. I am not sure how to source a large quality corpus of changes for the average case

Reply:
We do have cases in smd where we run b.Run within b.Run and it exactly does the right thing IMO:
https://github.com/kubernetes-sigs/structured-merge-diff/blob/9c99f49f6345f80f6653fc1e0fd827fd1a8bc651/typed/parser_test.go#L86

i.e. it creates a sub-benchmark, and indeed only measures the speed of the most nested layer. Still, you can organize your benchmarks into sub-benchmarks which I think would make this clearer.

Reply:
split the benchmark into Valid x Valid, Valid x Invalid, Invalid x Invalid yamls